lib/orderbook/OrderBook.ts

```diff
@@ -170,49 +192,120 @@ class OrderBook extends EventEmitter {
+  private addOwnOrder = async (
+    order: orders.StampedOwnOrder,
+    discardRemaining = false,
+    onUpdate?: (e: PlaceOrderEvent) => void,
+    maxTime?: number,
+  ): Promise<PlaceOrderResult> => {
+    // this method can be called recursively on swap failures retries.
+    // if max time exceeded, don't try to match
+    if (maxTime && Date.now() > maxTime) {
+      assert(discardRemaining, 'discardRemaining must be true on recursive calls where maxTime could exceed');
+      this.logger.info(`addOwnOrder max time exceeded. order (${JSON.stringify(order)}) won't be matched`);
+
+      // returning the remaining order to be rolled back and handled by the initial call
+      return Promise.resolve({
+        internalMatches: [],
+        swapResults: [],
+        remainingOrder: order,
+      });
     }
 
+    // fetch the current matchingEngine
     const matchingEngine = this.matchingEngines.get(order.pairId);
     if (!matchingEngine) {
       throw errors.PAIR_DOES_NOT_EXIST(order.pairId);
     }
 
-    const stampedOrder: orders.StampedOwnOrder = { ...order, id: uuidv1(), createdAt: ms() };
-    const matchingResult = matchingEngine.matchOrAddOwnOrder(stampedOrder, discardRemaining);
-    const { matches, remainingOrder } = matchingResult;
+    // perform match. maker orders will be removed from the repository
+    const matchingResult = matchingEngine.match(order);
 
-    if (matches.length > 0) {
-      matches.forEach(({ maker, taker }) => {
-        this.handleMatch({ maker, taker });
-      });
+    // instantiate the final response object
+    const result: PlaceOrderResult = {
+      internalMatches: [],
+      swapResults: [],
+      remainingOrder: matchingResult.remainingOrder,
+    };
+
+    // instantiate a container for failed swaps, for retry purposes
+    const swapFailures: StampedOwnOrder[] = [];
+
+    // append the taker quantity to the remaining order, after making sure it's initialized.
+    // if the maker is given, re-add it to the repository
+    const rejectNonInternalMatch = (taker: StampedOwnOrder, maker?: StampedPeerOrder) => {
+      result.remainingOrder = result.remainingOrder || { ...order, quantity: 0 };
+      result.remainingOrder.quantity += taker.quantity;
+
+      if (maker) {
+        matchingEngine.addPeerOrder(maker);
+      }
+    };
+
+    // iterate over the matches
+    for (const { maker, taker } of matchingResult.matches) {
+      const portion: OrderPortion = { orderId: maker.id, pairId: maker.pairId, quantity: maker.quantity };
+      if (orders.isOwnOrder(maker)) {
+        // internal match
+        result.internalMatches.push(maker);
+        this.emit('ownOrder.filled', portion);
+        onUpdate && onUpdate({ internalMatch: maker });
+      } else {
+        // non-internal match
+        if (!this.swaps) {
+          rejectNonInternalMatch(taker, maker);
+          continue;
+        }
+
+        this.emit('peerOrder.filled', portion);
```

Comment:
It might make sense to emit this event only after the swap succeeds. If it fails, maybe we call `removePeerOrder`?

Reply:
At the moment this logic isn't being handled properly, as we are missing swap error handling. 

The peer order was already removed in the matching stage. I added some logic for returning it in case swaps isn't available. We need to revise these section later on.

Reply:
Ok, that's fine but could you just add a TODO mentioning this in the code?